### ISSUE TITLE
The predictions by Command line and GUI are different

### ISSUE THREAD

**Issue by ycwei0321** [MouseLand/cellpose#972]
Hello,

I'm recently using the CellPose to do the cell count. However, I found the predictions between GUI and command line are very different that the GUI gave me a better prediction, can you give me some suggestions? 

Here is the command I used: 
cellpose --dir “D:/user/CellPose/input” --image_path “D:/user/CellPose/Input” --add_model “C:/Users/winshah/.cellpose/models/cyto2_cp3.npy” --chan 0 --use_gpu --flow_threshold 0.4 --cellprob_threshold 0 --save_rois --save_png --save_outlines --savedir “D:/user/CellPose/Output”

Here is the screenshot of the GUI I use:
![image](https://github.com/user-attachments/assets/f34aa791-e388-4f9f-8aac-25e559415b39)

If I want to get the similar result by command line, whant kind of parameter should I use?

I found if I use the pretrained model named cyto3, it gave me the opposite results. The result from command line is much better than the GUI.

Thanks for your help.

**Comment by ReallyCoolBean** [comment]
I also have the issue of different output in the GUI and from the command line using Cellpose2.0 and we are not alone, as I found others reporting that, however no good explanation or solution to the problem: https://github.com/MouseLand/cellpose/issues/843 and https://github.com/MouseLand/cellpose/issues/758

I trained a model using command line and now when I evaluate it through the command line in PyTorch, I have good results. Here is the code snippet:
![image](https://github.com/user-attachments/assets/534820fa-c48a-42ef-b30f-c22a4505520f)
and here is one of my images, a z-stack of dimensions: (41,2,2045,2046), here visualised in napari 15th slice in a stack with predicted masks. I am using 2 channels: green (2) as the main channel labelling cells and red (1) with nuclei.
![image](https://github.com/user-attachments/assets/228f4d02-6d32-418a-80e5-99d7d51de4ed)

Now this is what happens when I run the model with the same parameters in GUI:
![image](https://github.com/user-attachments/assets/234b3a17-1e45-4ae9-befe-f4e7a790888b)
Model does a poorer job, it predicts sometimes many tiny masks eg. top right corner instead of a one normal mask.

To make things even more confusing, when I try to run the same model on the same image in Arivis with the same parameters, this is what happens (again showing 15th slice from the z-stack). 
![image](https://github.com/user-attachments/assets/f03666aa-6bbf-4f7a-bb1c-3c33fd3586df)

Prediction is different than from Cellpose GUI and different than from the command line. Everything is run in the same environment. Does anyone have any idea what's going on? I don't know which result to trust. I have the best results fromt he command line and this is also where I did evaluation of different parameters and now I want to give the model to another person so they can work with it in Arivis and we can't replicate the results.

**Comment by carsen-stringer** [comment]
can you please check if this is the case in cellpose v3? this is the old GUI. also it would be helpful if you could please provide the image that has discrepancies in segmentation, then I can debug it, thanks

**Comment by ReallyCoolBean** [comment]
Hey, thank you for looking into this! I cloned our environment and upgraded to cellpose3 and the problem persists in the new GUI (screenshot attached). What is the best way to provide you with the image? It's 680MB, so I can't upload it here.
![image](https://github.com/user-attachments/assets/29c0ead8-f35e-462e-bc75-ac52e9dd4be5)

**Comment by carsen-stringer** [comment]
if you can provide a google drive link that would be great thanks, and the CLI command you are using

**Comment by ycwei0321** [comment]
> if you can provide a google drive link that would be great thanks, and the CLI command you are using

Hi

Thanks for your response. I can also provide some example images I tried before. Please see them in this drive link: https://drive.google.com/drive/folders/1UKnCFPk5G3RQkn65VJcfyzjkuvu7y-I4?usp=sharing

Below is the command I used for the cell count, I tried to keep the parameter same with the GUI.

cellpose --dir “D:/Yichao/CellPose/input” --image_path “D:/Yichao/CellPose/Input” --add_model “C:/Users/winshah/.cellpose/models/cyto2_cp3.npy” --chan 0 --use_gpu --flow_threshold 0.4 --cellprob_threshold 0 --save_rois --save_png --save_outlines --savedir “D:/Yichao/CellPose/Output”

If you have any questions, please let me know. Thanks for your kind help.

**Comment by carsen-stringer** [comment]
you've added ".npy" to the end of the model name, I think the command you want is 

```
python -m cellpose --verbose --dir /path/to/images/ --pretrained_model cyto2_cp3 --chan 0 --flow_threshold 0.4 --cellprob_threshold 0 --save_rois --save_png --save_outlines
```

with this command and using the "cyto2_cp3" model in the GUI I got the exact same result.

going to close this issue for now, but let me know if you have more questions

**Comment by ReallyCoolBean** [comment]
Here is a folder with my example image and the model I trained:
https://drive.google.com/drive/folders/1ptaGPKS77ihv1Ewwte1eVuVbX1r-_wPg?usp=sharing
Perhaps I didn't describe correctly what I'm doing, I am actually not using cellpose CLI, but cellpose API by running code in PyCharm. I was inspired by the colab notebook you posted here. Here is the line of code that I am using: 

masks = model.eval(test_data,
                   channels=[2, 1],
                   diameter=123,
                   cellprob_threshold=0,
                   flow_threshold=0.4,
                   stitch_threshold=0.2)[0]

where model is the model that I provided in the google drive and test_data is the image. The same model and the same image were loaded in GUI and I believe the parameters in GUI were also the same as in my code (see screenshots).

**Comment by carsen-stringer** [comment]
Since this is 3D, can you please check if the normalization parameters are the same in both cases? They are printed in the GUI as a dictionary, and you can input that dictionary into `normalize` in `eval`.

Regarding arivis outputs, their code is closed source and I don't have a license so we can't verify whether it would work the same way

**Comment by ReallyCoolBean** [comment]
Thank you for the advice! Sadly, I still get different results: GUI predicts 871 ROIs while the python script predicts 149. I ran again the model in Cellpose3 GUI and I got the same result as before (here showing plane 14):
![image](https://github.com/user-attachments/assets/87b7cce4-3208-4ca6-8044-dc24afd3d109)
I checked the output in the terminal and normalization parameters:
![image](https://github.com/user-attachments/assets/22093e09-f68e-4d60-9bf8-43bceb9729d9)
I then run the following python script:
![image](https://github.com/user-attachments/assets/9a2edb76-3742-4509-917b-9b73e0faf6e4)
and this is my result:
![image](https://github.com/user-attachments/assets/f3dc4ac7-20c8-497c-af1c-101af44666b7)

**Comment by carsen-stringer** [comment]
Can you include what is printed when the script runs? I think you're missing the do_3D=True flag in model.eval

**Comment by carsen-stringer** [comment]
Oh I see you are stitching, please do include what the script is printing

**Comment by ReallyCoolBean** [comment]
Well, it doesn't print much, just that images from my test set were loaded and that model finished finished running for the one image that I selected.
![image](https://github.com/user-attachments/assets/fc2f03ee-d39d-4faa-ae5c-240420191746)

**Comment by carsen-stringer** [comment]
hmm I am confused because in the GUI it looks like you have cell fragments that result from running the 2.5D model not the stitching model, are you sure that the screenshot is from running with stitch_threshold=0.2?

**Comment by ReallyCoolBean** [comment]
I loaded the image, set the settings like it's shown in the screenshot and then pressed 'run' and that's the result. I repeated that several times, on different days, just to make sure I didn't click something wrong by accident, and it's always the same result.
What is the 2.5D model?

**Comment by carsen-stringer** [comment]
the 2.5d model is described in the Cellpose1 paper, can you please post the command line info that prints when you run in the gui vs in the API?

**Comment by ReallyCoolBean** [comment]
So that's what I posted above, but I think the screenshot didn't capture everything from the GUI, so here it is again:
![image](https://github.com/user-attachments/assets/966b40ac-9fe2-4bc9-9fa6-e089d34447d8)
![image](https://github.com/user-attachments/assets/1db8e52d-01cd-4ce8-84cf-9531b64737de)
and from running my script:
![image](https://github.com/user-attachments/assets/3ba009d7-c8b2-4dc0-8bd5-7203689fd9f8)

**Comment by carsen-stringer** [comment]
thanks can you run the script with `io.logger_setup()` declared before you run cellpose so we can see the loaded model and the output of the script.

**Comment by carsen-stringer** [comment]
it looks like in the script you are running many images, and you may be comparing the output potentially from two different tiffs (indeed it should say 41 if it has 41 planes it loads in)

**Comment by HaofeiGe** [comment]
I have found the reason. I dont know why, but when I give up using "for”, instead of using "def+list", it works the same as it in GUI.
This is my code:
import tifffile
from cellpose import models, io, denoise
import os
from pathlib import Path

input_dir = Path('F:/温敏分子行为/D218/2024-10-31/Control_MIP/')
output_dir = Path('F:/温敏分子行为/D218/2024-10-31/Segmentation/')

output_dir.mkdir(parents=True, exist_ok=True)

model = models.Cellpose(model_type='nuclei', gpu=True)
dn_model = denoise.DenoiseModel(model_type="denoise_nuclei", gpu=True)

def process_file(file_path):
file_name = file_path.name
print(f"Processing file: {file_name}")
data = tifffile.imread(file_path)
print(f"Shape: {data.shape}")
print("Performing denoising...")
data_denoised = dn_model.eval(data, channels=None, do_3D=True, diameter=140)
print("Performing segmentation...")
masks, flows, styles, diams = model.eval(
data_denoised,
diameter=140,
channels=[0, 0],
normalize={
'lowhigh': None,
'percentile': [1.0, 99.0],
'normalize': True,
'norm3D': True,
'sharpen_radius': 0,
'smooth_radius': 0,
'tile_norm_blocksize': 0,
'tile_norm_smooth3D': 1,
'invert': False
},
flow_threshold=0.4,
cellprob_threshold=0,
do_3D=True,
anisotropy=2.5,
min_size=100
)
output_file_name = file_name.replace('.tiff', '_masks.tiff').replace('.tif', '_masks.tiff')
output_path = output_dir / output_file_name
io.save_masks(data_denoised, masks, flows, file_names=output_path, png=False, tif=True, channels=[0, 0])
print(f"Saved masks to {output_path}")

tiff_files = list(input_dir.glob('.tiff')) + list(input_dir.glob('.tif'))
list(map(process_file, tiff_files))

**Comment by carsen-stringer** [comment]
@ReallyCoolBean the GUI adds a channel for viewing in RGB. in your logger output from your script I don't see your custom model being loaded, is that the issue?

@HaofeiGe we'll discuss on #1026 since this is probably a different issue

**Comment by Jiadalee** [comment]
> [@ReallyCoolBean](https://github.com/ReallyCoolBean) the GUI adds a channel for viewing in RGB. in your logger output from your script I don't see your custom model being loaded, is that the issue?
> 
> [@HaofeiGe](https://github.com/HaofeiGe) we'll discuss on [#1026](https://github.com/MouseLand/cellpose/issues/1026) since this is probably a different issue

@ReallyCoolBean after reviewing your post, seems you were trying to compare results among cli, gui, and script runings.  I want to propose some points regarding your python script. As you are using GPU to train and eval model. firstly, pls make sure GPU is consistently set to be true for both training and eval stages since the CPU trained model has different outputs (masks shape/dims might be different) from the GPU trained model. You can play around the gpu settings to see the masks' differences.

Secondly, the images that you are working on has 2 channels and 3-dims. In that way, in your eval stage, you can try loop over each channel-wise pairs to pred the masks and then append the masks to how the results would be. As @carsen-stringer  mentioned, GUI uses 3 channels' pixels to train and eval model, your python script should use the same format.

Thirdly, someone at post #1026 mentioned GUI has better performance than command line or scrip run. It is not strictly true. Apart from the consistency of the parameter settings, we not only need to be aware of the computing source difference (CPU versus GPU), but also we need to understand how the images are fed into the model. Even with the same image datasets, the image feeding manner will cause differences in model performance. The feeding manner is very likely to sample different volumes of pixel datasets for the model training and evaluation. Someone feed the 3-channel RGB as one volume to the model, while others iteratively feed the channel-wise pairing configuration to the model. Someone feed the images plane by plane, while others feed the stacked planes to model.

**Comment by raeesahhy** [comment]
> 
> Thirdly, someone at post [#1026](https://github.com/MouseLand/cellpose/issues/1026) mentioned GUI has better performance than command line or scrip run. It is not strictly true. Apart from the consistency of the parameter settings, we not only need to be aware of the computing source difference (CPU versus GPU), but also we need to understand how the images are fed into the model. Even with the same image datasets, the image feeding manner will cause differences in model performance. The feeding manner is very likely to sample different volumes of pixel datasets for the model training and evaluation. Someone feed the 3-channel RGB as one volume to the model, while others iteratively feed the channel-wise pairing configuration to the model. Someone feed the images plane by plane, while others feed the stacked planes to model.
> 

Hi @Jiadalee could you please provide some guidance on how to feed 3D images into the CLI or a notebook in a way that mimics how they are fed into the GUI? This would be really helpful because I have over 200 3D images per timelapses dataset and I would like to have all my timelapses segmented. 

I have managed to successfully run CPSAM in a notebook for 2D images but with 3D images, I keep running into issues although I use the same parameters I use in the GUI. So if I'm not missing anything then it must be the image feeding issue you mentioned. 

For reference this is what I am finding works best for me in the GUI:

<img width="498" height="636" alt="Image" src="https://github.com/user-attachments/assets/66b26d6d-e7e2-477f-ba8a-ac5eb5b79dba" />

This is the CLI command I am using while trying to replicate the same result (but the segmentations that result are much less neat): 

`python -m cellpose --verbose --dir /Users/rhay0006/Documents/test2/images --use_gpu --pretrained_model cpsam --chan 0 --flow_threshold 0.8 --cellprob_threshold -1 --stitch_threshold 0.2  --do_3D --z_axis 0 --batch_size 32 --niter 200  --save_rois --save_tif --save_outlines `

Thanks.
